mongomock/store.py

```diff
@@ -149,10 +149,13 @@ def _expire_documents(self, index):
         # "key" structure = list of (field name, direction) tuples
         ttl_field_name = index['key'][0][0]
         ttl_now = mongomock.utcnow()
-        expired_ids = [
-            doc['_id'] for doc in six.itervalues(self._documents)
-            if self._value_meets_expiry(doc.get(ttl_field_name), expiry, ttl_now)
-        ]
+
+        # 694 - Lock the whole iteration to avoid RuntimeError when mutating OrderedDicts
+        with lock:
```

Comment:
You would also need to lock anywhere iterating over documents, especially the @property document above.

I think a multiple read lock approach would be the right fix instead.
https://code.activestate.com/recipes/577803-reader-writer-lock-with-priority-for-writers/

Reply:
Thanks @pcorpet. I'll try implementing it soon! :)

Reply:
@pcorpet I've added the plain implementation with just some context managers as helpers. Tried to do a wrap around OrderedDict but it was 2x slower it seems.